File: sar_project/Analytical_Optical_Flow.cpp
#include <iostream>
#include <Eigen/Dense>
#include <opencv2/opencv.hpp>  // Include OpenCV for image handling
#include <fstream>  // Include this for file I/O

// Function to save Eigen matrix as CSV
void saveMatrixAsCSV(const Eigen::MatrixXd& matrix, const std::string& filename) {
    std::ofstream file(filename);
    if (file.is_open()) {
        for (int i = 0; i < matrix.rows(); ++i) {
            for (int j = 0; j < matrix.cols(); ++j) {
                file << std::fixed << std::setprecision(2) << matrix(i, j);  // Format to 2 decimal places
                if (j != matrix.cols() - 1) {
                    file << ",";  // Add a comma between elements
                }
            }
            file << "\n";  // New line at the end of each row
        }
        file.close();
        std::cout << "Matrix saved as " << filename << std::endl;
    } else {
        std::cerr << "Error: Could not open file " << filename << std::endl;
    }
}

// Function to save Eigen vector as CSV
void saveVectorAsCSV(const Eigen::VectorXd& vector, const std::string& filename) {
    std::ofstream file(filename);
    if (file.is_open()) {
        for (int i = 0; i < vector.size(); ++i) {
            file << std::fixed << std::setprecision(2) << vector(i);  // Format to 2 decimal places
            if (i != vector.size() - 1) {
                file << ",";  // Add a comma between elements
            }
        }
        file << "\n";  // New line at the end
        file.close();
        std::cout << "Vector saved as " << filename << std::endl;
    } else {
        std::cerr << "Error: Could not open file " << filename << std::endl;
    }
}

int main() {

    // Load the current and previous images using OpenCV
    std::string cur_img_path = "/home/dlee/ros2_ws/src//sar_simulation/sar_project/Generated_Cur_img_1.png";
    std::string prev_img_path = "/home/dlee/ros2_ws/src//sar_simulation/sar_project/Generated_Prev_img_1.png";
    
    // Load images as grayscale (single channel)
    cv::Mat cur_img_cv = cv::imread(cur_img_path, cv::IMREAD_GRAYSCALE);
    cv::Mat prev_img_cv = cv::imread(prev_img_path, cv::IMREAD_GRAYSCALE);

    // Check if images are loaded successfully
    if (cur_img_cv.empty()) {
        std::cerr << "Error: Could not load current image at " << cur_img_path << std::endl;
        return -1;
    }
    if (prev_img_cv.empty()) {
        std::cerr << "Error: Could not load previous image at " << prev_img_path << std::endl;
        return -1;
    }

    // Resize the images to 32x32 pixels if needed
    cv::resize(cur_img_cv, cur_img_cv, cv::Size(32, 32));
    cv::resize(prev_img_cv, prev_img_cv, cv::Size(32, 32));

    int HEIGHT_PIXELS = cur_img_cv.rows;
    int WIDTH_PIXELS = cur_img_cv.cols;
    int O_up = WIDTH_PIXELS / 2;
    int O_vp = HEIGHT_PIXELS / 2;

    // Convert OpenCV Mat to Eigen Matrix for cur_img and prev_img
    Eigen::MatrixXd cur_img(HEIGHT_PIXELS, WIDTH_PIXELS);
    Eigen::MatrixXd prev_img(HEIGHT_PIXELS, WIDTH_PIXELS);

    for (int v = 0; v < HEIGHT_PIXELS; ++v) {
        for (int u = 0; u < WIDTH_PIXELS; ++u) {
            cur_img(v, u) = static_cast<double>(cur_img_cv.at<uchar>(v, u));
            prev_img(v, u) = static_cast<double>(prev_img_cv.at<uchar>(v, u));
        }
    }

    // Save cur_img as a CSV file
    saveMatrixAsCSV(cur_img, "/home/dlee/ros2_ws/src/sar_simulation/sar_project/cur_img.csv");
    saveMatrixAsCSV(prev_img, "/home/dlee/ros2_ws/src/sar_simulation/sar_project/prev_img.csv");

    // Additional code for the image processing
    int I_0 = 255; //Brightness value (0-255)
    float FoV = 82.22; //Field of View [deg]

    float w = 3.6e-6;
    float f = 0.066e-3;
    float delta_t = 0.01667;

    Eigen::Matrix<double, 1, 3> Ku_1;
    Ku_1 << -1, 0, 1;
    Eigen::Matrix<double, 3, 1> Ku_2;
    Ku_2 << 1, 2, 1;

    Eigen::Matrix<double, 1, 3> Kv_1;
    Kv_1 << 1, 2, 1;
    Eigen::Matrix<double, 3, 1> Kv_2;
    Kv_2 << -1, 0, 1;

    Eigen::MatrixXd G_up = Eigen::MatrixXd::Zero(HEIGHT_PIXELS, WIDTH_PIXELS);
    Eigen::MatrixXd G_vp = Eigen::MatrixXd::Zero(HEIGHT_PIXELS, WIDTH_PIXELS);
    Eigen::MatrixXd G_rp = Eigen::MatrixXd::Zero(HEIGHT_PIXELS, WIDTH_PIXELS);
    Eigen::MatrixXd G_tp = Eigen::MatrixXd::Zero(HEIGHT_PIXELS, WIDTH_PIXELS);

    // Calculate image gradients
    for (int v_p = 1; v_p < HEIGHT_PIXELS - 1; ++v_p) {
        for (int u_p = 1; u_p < WIDTH_PIXELS - 1; ++u_p) {
            Eigen::MatrixXd patch = cur_img.block<3, 3>(v_p - 1, u_p - 1);

            G_vp(v_p, u_p) = (Ku_1 * patch).dot(Ku_2);
            G_up(v_p, u_p) = (Kv_1 * patch).dot(Kv_2);

            G_rp(v_p, u_p) = (2 * (u_p - O_up) + 1) * G_up(v_p, u_p) + (2 * (v_p - O_vp) + 1) * G_vp(v_p, u_p);
            G_tp(v_p, u_p) = cur_img(v_p, u_p) - prev_img(v_p, u_p);
        }
    }
    saveMatrixAsCSV(G_up, "/home/dlee/ros2_ws/src/sar_simulation/sar_project/G_up_matrix.csv");
    saveMatrixAsCSV(G_vp, "/home/dlee/ros2_ws/src/sar_simulation/sar_project/G_vp_matrix.csv");
    saveMatrixAsCSV(G_rp, "/home/dlee/ros2_ws/src/sar_simulation/sar_project/G_rp_matrix.csv");
    saveMatrixAsCSV(G_tp, "/home/dlee/ros2_ws/src/sar_simulation/sar_project/G_tp_matrix.csv");

    Eigen::MatrixXd X(3, 3);
    X(0, 0) = f * G_vp.array().square().sum();
    X(0, 1) = -f * (G_up.array() * G_vp.array()).sum();
    X(0, 2) = -(w / 2) * (G_rp.array() * G_vp.array()).sum();

    X(1, 0) = f * (G_vp.array() * G_up.array()).sum();
    X(1, 1) = -f * G_up.array().square().sum();
    X(1, 2) = -(w / 2) * (G_rp.array() * G_up.array()).sum();

    X(2, 0) = f * (G_vp.array() * G_rp.array()).sum();
    X(2, 1) = -f * (G_up.array() * G_rp.array()).sum();
    X(2, 2) = -(w / 2) * G_rp.array().square().sum();

    Eigen::Vector3d y;
    y(0) = (G_tp.array() * G_vp.array()).sum();
    y(1) = (G_tp.array() * G_up.array()).sum();
    y(2) = (G_tp.array() * G_rp.array()).sum();
    y *= (8 * w / delta_t);

    // Use JacobiSVD to calculate pseudo-inverse manually
    Eigen::JacobiSVD<Eigen::MatrixXd> svd(X, Eigen::ComputeThinU | Eigen::ComputeThinV);
    Eigen::MatrixXd S = svd.singularValues().asDiagonal();
    Eigen::MatrixXd S_inv = S;
    for (int i = 0; i < S.rows(); ++i) {
        if (S(i, i) > 1e-6) {
            S_inv(i, i) = 1.0 / S(i, i);  // Inverse of non-zero singular values
        } else {
            S_inv(i, i) = 0;  // Zero out small singular values to avoid instability
        }
    }

    Eigen::MatrixXd pinv_X = svd.matrixV() * S_inv * svd.matrixU().transpose();  // Pseudo-inverse calculation

    Eigen::Vector3d b = pinv_X * y;

    // Output the solution vector b
    std::cout << "Solution vector b: " << b.transpose() << std::endl;
    //std::cout << "Solution vector b: " << 1/b[2] << std::endl;

    // Save X matrix as a CSV file
    saveMatrixAsCSV(X, "/home/dlee/ros2_ws/src/sar_simulation/sar_project/X_matrix.csv");

    // Save y vector as a CSV file
    saveVectorAsCSV(y, "/home/dlee/ros2_ws/src/sar_simulation/sar_project/y_vector.csv");

    return 0;
}

File: sar_projects_1/Analytical_Optical_Flow.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.animation as animation
from matplotlib.patches import Rectangle
import os
import cv2

## PLOT BRIGHTNESS PATTERN FROM 2.4.1 HORIZONTAL MOTION
I_0 = 255   # Brightness value (0-255)

## CAMERA PARAMETERS
FoV = 82.22 # Field of View [deg]
FoV_rad = np.radians(FoV) # Field of View [deg]

WIDTH_PIXELS = 32
HEIGHT_PIXELS = 32
w = 3.6e-6              # Pixel width [m]
f = 0.066e-3             # Focal Length [m]
O_up = WIDTH_PIXELS/2   # Pixel X_offset [pixels]
O_vp = HEIGHT_PIXELS/2  # Pixel Y_offset [pixels]


class Optical_Flow():

    def __init__(self,L,FPS):
        """_summary_

        Args:
            L (float):  [m/s]
            FPS (uint): Frames per second
        """              

        self.L = L*2 # Double the Feature width for sine period
        self.FPS = FPS

    def Generate_Camera_Image(self,D_cam=1,X_cam=0):

        U_p,V_p = np.meshgrid(np.arange(0,WIDTH_PIXELS,1),np.arange(0,HEIGHT_PIXELS,1))

        ## CONVERT PIXEL INDEX TO METERS
        u =  (U_p - O_up)*w + w/2 
        v =  (U_p - O_up)*w + w/2

        ## RETURN AVERAGE BRIGHTNESS VALUE OVER PIXEL
        I = I_0/2 * (f*self.L/(np.pi*D_cam*w) * (np.sin(np.pi*D_cam*w/(self.L*f))*np.sin(2*np.pi/self.L * (D_cam*u/f + X_cam))) + 1)

        return I

    def Generate_Pattern(self,Surf_width=2,Surf_Height=2,pixel_density=100,save_img=False,X_cam=0.0,D_cam=0.5):
        """Save show pattern and save image to file. As well, display camera boundaries for validation

        Args:
            Surf_width (int, optional): Width of pattern to preserve pixel density. Defaults to 2.
            Surf_Height (int, optional): Heigt of pattern to preserve pixel density. Defaults to 2.
            pixel_density (int, optional): Number of pixels per meter. Defaults to 100.
            save_img (bool, optional): Save image to file. Defaults to False.
            X_cam (float, optional): X location of camera. Defaults to 0.5.
            D_cam (float, optional): Distance of camera from pattern surface. Defaults to 0.5.
        """        

        ## GENERATE PATTERN BOUNDS
        x = np.linspace(-0.5*Surf_width, 0.5*Surf_width, pixel_density*Surf_width)
        y = np.linspace(-0.5*Surf_Height,0.5*Surf_Height,pixel_density*Surf_Height)
        X,Y = np.meshgrid(x,y)

        ## GENERATE PATTERN
        I = I_0/2*(np.sin(2*np.pi/self.L*X) + 1)
        # I = np.where(I < 128,0,255)

        ## GENERATE CAMERA BOUNDS
        Img_Width = 2*np.tan(FoV_rad/2)*D_cam
        Img_Height = 2*np.tan(FoV_rad/2)*D_cam

        
        ## CREATE FIGURE OF PATTERN
        fig = plt.figure()
        ax = fig.add_subplot(111)

        ## PLOT PATTERN AND CAMERA BOUNDARY
        ax.imshow(I, interpolation='none', 
            vmin=0, vmax=255, cmap=cm.gray,
            origin='upper',
            extent=[x.min(),x.max(),y.min(),y.max()]
            )

        ax.add_patch(
            Rectangle(
                (X_cam-Img_Width/2,-Img_Height/2),
                Img_Width,
                Img_Height,
                lw=1,fill=False,color="tab:blue")
            )

        ## SHOW PLOT
        plt.show()

        ## SAVE PATTERN TO FILE
        if save_img == True:
            plt.imsave(
                f'/home/dlee/Pictures/project/Strip_Pattern_W_{Surf_width}-H_{Surf_Height}.png', 
                I, 
                vmin=0, vmax=255, cmap=cm.gray,
                origin='upper',
            )


    def Optical_Flow_Traj(self,D_0,Vz,X_0,Vx):

        Tau_List = []
        Tau_est_List = []
        Theta_x_List = []
        Theta_x_est_List = []
        t_List = []

        D_perp_prev = D_0
        X_cam_prev = X_0

        t_max = 1

        ## IMAGE SENSOR PLOT
        fig = plt.figure()
        ax = fig.add_subplot(111)
        im = ax.imshow(self.Generate_Camera_Image(D_cam=D_0,X_cam=X_0), interpolation='none', 
                        vmin=0, vmax=255, cmap=cm.gray,
                        origin='upper',
        )
        ax.set_title("Image Sensor Pattern (Pixels)")
        ax.set_xlabel("u_p [pixels]")
        ax.set_ylabel("v_v [pixels]")
        fig.tight_layout()

        def animate_func(i):

            ## UPDATE CAMERA POSITION
            t = i/self.FPS   # Time [s]

            D_perp = D_0 - Vz*t    # Distance to Ceiling [m]
            X_cam = X_0 + Vx*t

            D_perp_prev = D_0 - Vz*(t-1/self.FPS)
            X_cam = X_0 + Vx*(t-1/self.FPS)


            Tau = D_perp/(Vz+1e-6)
            Theta_x = Vx/D_perp

            ## STOP ANIMATION IF PAST CEILING SURFACE
            if D_perp <= 0.01:
                return
            
            ## CALCULATE OPTICAL FLOW ESTIMATES
            Prev_img = self.Generate_Camera_Image(D_cam=D_perp_prev,X_cam=X_cam_prev)
            Cur_img = self.Generate_Camera_Image(D_cam=D_perp,X_cam=X_cam)
            b = self.OF_Calc_Opt_Sep(Cur_img,Prev_img,1/self.FPS)
            Theta_x_est,Theta_y_est,Theta_z_est = b
            Tau_est = 1/Theta_z_est
            
            print(f"Tau: {Tau:.3f} | Tau_est: {Tau_est:.3f}")
            print(f"Theta_x: {Theta_x:.3f} | Theta_x_est: {Theta_y_est:.3f}\n")


            ## APPEN OPTICAL FLOW ESTIMATES TO LIST FOR PLOTTING
            Tau_List.append(Tau)
            Theta_x_List.append(Theta_x)

            Tau_est_List.append(Tau_est)
            Theta_x_est_List.append(-Theta_y_est)
       
            t_List.append(t)

            ## UPDATE IMAGE
            im.set_array(Cur_img)
            
        anim = animation.FuncAnimation(fig, 
                                    animate_func, 
                                    frames = t_max * self.FPS,
                                    interval = 1000 / self.FPS, # in ms
                                    repeat = False
                                    )

        plt.show()

        ## TAU PLOT
        fig2 = plt.figure(1)
        ax = fig2.add_subplot(111)

        ax.plot(t_List,Tau_est_List,'rx',label="Tau_estimate")
        ax.plot(t_List,Tau_List,label="Tau_actual")
        # ax.set_title(f'Tau Estimation - V: {vel:.2f} | Phi: {phi:.2f} | L: {L:.3f}')
        ax.set_xlabel('Time [s]')
        ax.set_ylabel('Tau [s]')
        ax.grid()
        ax.legend()

        fig2.tight_layout()

    def OF_Calc_Opt_Sep(self,cur_img,prev_img,delta_t):

        """Calculate optical flow values with seperable convolution and integer optimizations.
        Derivation in (Research_Notes_Book_2.pdf)

        Args:
            cur_img (np.array): Array of current image
            prev_img (np.array): Array of previous image
            delta_t (float): Time between images

        Returns:
            np.array: Array of (Theta_x_est,Theta_y_est,Theta_z_est)
        """        

        ## SEPERATED SOBEL KERNAL (U--DIRECTION)
        Ku_1 = np.array([[-1,0,1]]).reshape(3,1)
        Ku_2 = np.array([[ 1,2,1]]).reshape(1,3)


        ## SEPERATED SOBEL KERNAL (V--DIRECTION)
        Kv_1 = np.array([[ 1,2,1]]).reshape(3,1)
        Kv_2 = np.array([[-1,0,1]]).reshape(1,3)


        ## PRE-ALLOCATE INTENSITY GRADIENTS
        G_up = np.zeros((HEIGHT_PIXELS,WIDTH_PIXELS))
        G_vp = np.zeros((HEIGHT_PIXELS,WIDTH_PIXELS))
        G_rp = np.zeros((HEIGHT_PIXELS,WIDTH_PIXELS))
        G_tp = np.zeros((HEIGHT_PIXELS,WIDTH_PIXELS))


        ## CALCULATE IMAGE GRADIENTS
        for v_p in range(1, HEIGHT_PIXELS-1): 
            for u_p in range(1, WIDTH_PIXELS-1):
                G_up[v_p,u_p] = (Ku_2.dot((cur_img[v_p-1:v_p+2,u_p-1:u_p+2].dot(Ku_1)))).item()
                G_vp[v_p,u_p] = (Kv_2.dot((cur_img[v_p-1:v_p+2,u_p-1:u_p+2].dot(Kv_1)))).item()
                G_rp[v_p,u_p] = (2*(u_p - O_up) + 1)*G_up[v_p,u_p] + (2*(v_p - O_vp) + 1)*G_vp[v_p,u_p]
                G_tp[v_p,u_p] = cur_img[v_p,u_p] - prev_img[v_p,u_p]

        np.savetxt('/home/dlee/Pictures/project/G_up_matrix.csv', G_up, delimiter=',', fmt='%.6f')
        np.savetxt('/home/dlee/Pictures/project/G_vp_matrix.csv', G_vp, delimiter=',', fmt='%.6f')
        np.savetxt('/home/dlee/Pictures/project/G_rp_matrix.csv', G_rp, delimiter=',', fmt='%.6f')
        np.savetxt('/home/dlee/Pictures/project/G_tp_matrix.csv', G_tp, delimiter=',', fmt='%.6f')

        ## SOLVE LEAST SQUARES PROBLEM
        X = np.array([
            [f*np.sum(G_vp*G_vp), -f*np.sum(G_up*G_vp), -w/2*np.sum(G_rp*G_vp)],
            [f*np.sum(G_vp*G_up), -f*np.sum(G_up*G_up), -w/2*np.sum(G_rp*G_up)],
            [f*np.sum(G_vp*G_rp), -f*np.sum(G_up*G_rp), -w/2*np.sum(G_rp*G_rp)]
        ])

        y = np.array([
            [np.sum(G_tp*G_vp)],
            [np.sum(G_tp*G_up)],
            [np.sum(G_tp*G_rp)]
        ])*(8*w/delta_t)

        np.savetxt('/home/dlee/Pictures/project/X_matrix.csv', X, delimiter=',', fmt='%.6f')
        np.savetxt('/home/dlee/Pictures/project/y_vector.csv', y, delimiter=',', fmt='%.6f')

        ## SOLVE b VIA PSEUDO-INVERSE
        b = np.linalg.pinv(X)@y
        b = b.flatten()

        return b

if __name__ == '__main__':

    OF = Optical_Flow(L=0.25,FPS=60)
    #OF.Generate_Pattern(D_cam=1.0,Surf_width=4,Surf_Height=4,save_img=True)
    Cur_img_1 = OF.Generate_Camera_Image(D_cam=0.9, X_cam=0)
    plt.imsave('/home/dlee/Pictures/project/Generated_Cur_img_1.png', Cur_img_1, cmap=cm.gray, vmin=0, vmax=255)
    np.savetxt('/home/dlee/Pictures/project/Generated_Cur_img_1.csv', Cur_img_1, delimiter=',', fmt='%.2f')

    Prev_img_1 = OF.Generate_Camera_Image(D_cam=1, X_cam=0)
    plt.imsave('/home/dlee/Pictures/project/Generated_Prev_img_1.png', Prev_img_1, cmap=cm.gray, vmin=0, vmax=255)
    np.savetxt('/home/dlee/Pictures/project/Generated_Prev_img_1.csv', Prev_img_1, delimiter=',', fmt='%.2f')

    cur_img_path = '/home/dlee/Pictures/project/Generated_Cur_img_1.png'
    cur_img_cv = cv2.imread(cur_img_path, cv2.IMREAD_GRAYSCALE)
    if cur_img_cv is not None:
        cur_img_cv = cur_img_cv.astype(np.float64) 
        np.savetxt('/home/dlee/Pictures/project/Loaded_Cur_img_1.csv', cur_img_cv, delimiter=',', fmt='%.2f')
        print("Current image loaded and saved as CSV successfully.")
    else:
        print(f"Error loading current image: {cur_img_path}")

    # Load the previous image
    prev_img_path = '/home/dlee/Pictures/project/Generated_Prev_img_1.png'
    prev_img_cv = cv2.imread(prev_img_path, cv2.IMREAD_GRAYSCALE)
    if prev_img_cv is not None:
        prev_img_cv = prev_img_cv.astype(np.float64)
        np.savetxt('/home/dlee/Pictures/project/Loaded_Prev_img_1.csv', prev_img_cv, delimiter=',', fmt='%.2f')
        print("Previous image loaded and saved as CSV successfully.")
    else:
        print(f"Error loading previous image: {prev_img_path}")

    
    #np.savetxt('/home/dlee/Pictures/project/X_matrix.csv', X, delimiter=',', fmt='%.6f')
    #np.savetxt('/home/dlee/Pictures/project/y_vector.csv', y, delimiter=',', fmt='%.6f')

    b = OF.OF_Calc_Opt_Sep(cur_img_cv,prev_img_cv,1/60)
    Theta_x_est,Theta_y_est,Theta_z_est = b
    Tau_est = 1/Theta_z_est
    print()
    


    OF.Optical_Flow_Traj(X_0=0,Vx=0,D_0=1.0,Vz=1.0)

    #anim = OF.Optical_Flow_Traj(X_0=0,Vx=0.0,D_0=0.5,Vz=0.1)

    #plt.show()